File: src/dartlab/gather/mixins/news.py
"""Gather news/dartDoc mixin — 뉴스 RSS + DART 공시 원문 2 메서드."""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import polars as pl

from ..infra.http import runAsync
from ..infra.telemetry import emitGatherFetch
from ..sources import naverNews as _naver
from ..sources import news as _news
from .context import GatherMixinContext


class _GatherNewsMixin(GatherMixinContext):
    """뉴스 + DART 공시 원문 메서드 모음 — Gather 클래스 2 메서드."""

    def news(self, query: str, *, market: str = "KR", days: int = 30) -> "pl.DataFrame":
        """뉴스 검색 — KR=네이버 우선(+Google 폴백), 그 외 Google News RSS.

        Capabilities:
            - KR: 네이버 검색 API 우선 (제목+스니펫). 키 미설정/빈결과 시 Google News RSS 폴백.
            - KR 외: Google News RSS
            - circuit breaker + TTL 캐시 (DARTLAB_TTL_NEWS override)
            - DataFrame: date, title, source, url, description 컬럼

        AIContext:
            - sentiment / event-driven 분석의 외부 신호 원천
            - 본문은 untrusted external — provider 가 [EXTERNAL CONTENT START] 마커 적용
            - 소스는 내부 라우팅 (반환 source 컬럼이 naver/google_news 구분)

        Guide:
            query 는 자유 문자열 (종목명 + 키워드 조합 가능). 네이버 결과는 라이브
            표시용 (비영속) — 공개 적재는 별도 private 경로 (naverNews archive).

        When:
            sentiment / event 분석 / catalyst 모니터링 시.

        How:
            KR → naverNews._fetchAsync → (빈결과) news._fetchAsync(google) → toDataFrame.

        Args:
            query: 검색어 (종목명, 키워드 등).
            market: "KR" 또는 "US". 기본 "KR".
            days: 최근 N일 뉴스 (Google 경로). 기본 30.

        Returns:
            pl.DataFrame — date, title, source, url, description 컬럼.
            결과 없으면 빈 DataFrame.

        Requires:
            없음 (네이버 키 없으면 자동 Google 폴백 — 기존 동작 보존).

        Raises:
            없음 — fetch 실패는 빈 DataFrame.

        Example::

            g = getDefaultGather()
            g.news("삼성전자")                # KR — 네이버 우선
            g.news("Apple", market="US")     # US — Google News
            g.news("반도체", days=7)          # 최근 7일

        See Also:
            ``dartDoc`` — DART 공시 본문 (RSS 가 아닌 viewer 직접).
        """
        t0 = time.monotonic()
        cacheHit = False
        try:
            cache_key = f"{query}:{market}:news"
            cached = self._cache.getTyped(cache_key, "news")
            if cached is not None:
                cacheHit = True
                return cached  # type: ignore[return-value]
            if market.upper() == "KR":
                items = runAsync(_naver._fetchAsync(query, market=market, client=self._client))
                if not items:
                    items = runAsync(_news._fetchAsync(query, market=market, days=days, client=self._client))
            else:
                items = runAsync(_news._fetchAsync(query, market=market, days=days, client=self._client))
            df = _news.toDataFrame(items)
            if not df.is_empty():
                self._cache.putTyped(cache_key, "news", df)
            return df
        finally:
            emitGatherFetch("news", (time.monotonic() - t0) * 1000, cacheHit=cacheHit, market=market)

    def dartDoc(self, rceptNo: str) -> "pl.DataFrame":
        """DART 공시 viewer 원문 fetch (14자리 rcept_no, 무인증).

        Capabilities:
            - 14자리 rcept_no 만으로 공시 원문 fetch
            - viewer 인덱스 페이지 (dsaf001/main.do) sub-doc 목차 파싱
            - 각 섹션 HTML → 텍스트 (테이블 마크다운 보존)
            - API key 불필요 (providers/dart/openapi 와 분리)
            - rate limit + circuit breaker + TTL 캐시 (gather/infra 자동)

        Args:
            rceptNo: 14자리 접수번호 (예: "20240315000123").

        Returns:
            pl.DataFrame — section_order, title, url, text 컬럼.

        Requires:
            네트워크 (dart viewer 무인증 접근).

        Raises:
            InvalidRceptNoError: rceptNo 가 14자리 숫자 아님.
            DocumentNotFoundError: viewer 가 sub-doc 을 반환하지 않음.

        Untrusted Body:
            viewer 본문은 외부 1차 출처지만 AI 엔진 소비 시
            ``Ref.sourceType="external"`` 마커로 감싸야 한다 (호출자 책임).

        Example::

            g = getDefaultGather()
            df = g.dartDoc("20240315000123")
        """
        t0 = time.monotonic()
        try:
            from dartlab.gather.dart.viewer import fetch as _fetchDartDoc

            return _fetchDartDoc(rceptNo, client=self._client)
        finally:
            emitGatherFetch("dartDoc", (time.monotonic() - t0) * 1000, cacheHit=False, market="KR")

File: ui/packages/contracts/src/news.ts
// 뉴스 계약 — 종목별 뉴스 헤드라인(제목+스니펫+원문링크). 터미널 우측패널·주가 이벤트레일 소비.
// 빈값 규약(runtime.ts): [] = 조회 성공·해당 없음(예: 시총 상위 외 종목·미배선·프록시 미설정).
// 출처: 네이버 검색 API archive(언론사 저작권·private). 어댑터는 워커 서버사이드 read 로만 표시(재배포 아님).

/** 뉴스 트랙 — naver=검색 API(스니펫 O) · gdelt=GDELT DOC API(제목+링크) · google=Google News RSS 라이브
 *  (조회시점 최신 헤드라인, 스니펫 없음). 좌우 분리: gdelt=우(과거), 그 외(naver·google)=좌(최근). */
export type NewsTrack = 'naver' | 'gdelt' | 'google';

export interface NewsItem {
	date: string; // YYYY-MM-DD (발행일)
	title: string; // 기사 제목
	source: string; // 언론사/도메인
	url: string; // 원문 링크 (클릭 시 외부 이동)
	description: string; // 스니펫(요약) — gdelt 는 빈 문자열
	track: NewsTrack; // 소스 트랙 (좌우 분리 키)
}

// 시장 전체(cross) 뉴스 헤드라인 — 좌측 터미널 패널. 종목별(NewsItem, 우측패널)과 정반대 멘탈모델:
// 종목 무관·전 시장 시간순·제목+원문링크만(스니펫 없음·stock_code 없음). 클릭=외부 기사 이동.
// 출처: public rss 아카이브(Google News RSS·재배포 가능). naver(저작권 private)는 우측 라이브 read 전용.
export interface MarketNews {
	date: string; // YYYY-MM-DD (발행일)
	title: string; // 기사 제목
	source: string; // 언론사/도메인
	url: string; // 원문 링크 (클릭 시 외부 이동)
}

export interface NewsPort {
	/** 종목별 최근 뉴스 (date 내림차순). 해당 없음·미배선·프록시 미설정은 [].
	 *  name(회사명) 주입 시 워커가 Google News RSS 라이브 헤드라인을 byCompany 아카이브 위에 머지(조회시점 최신). */
	forCompany(code: string, name?: string): Promise<NewsItem[]>;
	/** 시장 전체 최근 뉴스 (date 내림차순, 제목+링크). 좌측 cross 피드. 미배선·실패는 []. */
	market(): Promise<MarketNews[]>;
}
